go-controller/pkg/ovn/gress_policy.go

```diff
@@ -266,12 +281,17 @@ func (gp *gressPolicy) getMatchFromIPBlock(lportMatch, l4Match string) []string
 	return ipBlockMatches
 }
 
-// addNamespaceAddressSet adds a namespace address set to the gress policy
-// if the address set does not exist and returns `true`;  if the address set already exists,
-// it returns `false`.
+// addNamespaceAddressSet adds a namespace address set to the gress policy.
+// If the address set is not found in the db, return error.
+// If the address set is already added for this policy, return false, otherwise returns true.
 // This function is safe for concurrent use, doesn't require additional synchronization
-func (gp *gressPolicy) addNamespaceAddressSet(name string) bool {
-	v4HashName, v6HashName := addressset.MakeAddressSetHashNames(name)
+func (gp *gressPolicy) addNamespaceAddressSet(name string, asf addressset.AddressSetFactory) (bool, error) {
+	dbIDs := getNamespaceAddrSetDbIDs(name, gp.controllerName)
+	as, err := asf.EnsureAddressSet(dbIDs)
```

Comment:
The change itself was motivated by the added visibility of the fact that network policy uses address sets it doesn't create, so we need to check if the address set exists first, otherwise we'll get ovn-controller errors. That is also the difference with `delNamespaceAddressSet`, because we can safely delete address set that e.g. never existed, no additional checks are needed. I am not sure how to express this better in the naming.signatures, maybe you have some ideas?
So this change is kinda related to the new indexing in my mind, but I can do the intermediate commit with `GetHashNamesForAS` if you'd like that more :)

Reply:
But `EnsureAddressSet` it's not just `check if the address set exists first`, it actually creates the address set if it doesn't.
Do we need to protect against any potential race condition with the namespace handler? If we read here first to check it exists and it doesn't, and before we create it the namespace handler creates it, we could potentially overwrite. 

Usually we would hold the namespace lock for this. I would leave a comment explaining why if it is not needed.

A good name could be `ensureNamespaceAddressSet`, breaks symmetry with the delete name. I would explain in the godoc text what it does as well.